src/manifold/src/impl.cpp

```diff
@@ -655,25 +661,24 @@ void Manifold::Impl::IncrementMeshIDs(int start, int length) {
   ASSERT(start >= 0 && length >= 0 && start + length <= triBary.size(),
          logicErr, "out of bounds");
   const auto policy = autoPolicy(length);
-  // Use double the space since the Boolean has P and Q instances.
-  VecDH<int> includesMeshID(2 * Manifold::Impl::meshIDCounter_, 0);
+  HashTable<uint32_t> meshidTable(std::max(16u, meshids * 2));
 
   auto begin = triBary.begin() + start;
   auto end = begin + length;
-
-  for_each(policy, begin, end, MarkMeshID({includesMeshID.ptrD()}));
-
-  inclusive_scan(autoPolicy(includesMeshID.size()), includesMeshID.begin(),
-                 includesMeshID.end(), includesMeshID.begin());
-
-  const int numMeshIDs = includesMeshID[includesMeshID.size() - 1];
+  while (1) {
+    for_each(policy, begin, end, MarkMeshID({meshidTable.D()}));
+    if (!meshidTable.Full()) break;
+    meshidTable = HashTable<uint32_t>(meshidTable.Size() * 2);
+  }
+  inclusive_scan(
+      autoPolicy(meshidTable.Size()), meshidTable.GetValueStore().begin(),
+      meshidTable.GetValueStore().end(), meshidTable.GetValueStore().begin());
+  const int numMeshIDs = meshidTable.GetValueStore().back();
```

Comment:
If you just want the number of hash table entires, can you just call `Entries()`?

Reply:
Ah true!

Reply:
Oh it doesn't work, because every value in the map will be 1.

Reply:
The problem in the original issue is that, the vector being allocated has size the global mesh id counter, which can be very large even when the mesh object itself only consists of a few mesh ids. A hash table is used so we can ignore those meshids that are not present in the current mesh.
For performance, I did not measure it against the ols version. It may be a bit slower in general, but as the global counter keep increasing, it will eventually become faster than the original one and uses less memory.

Reply:
Ah, thank you, I missed what `meshids` was doing. Now that this is a hash table, is this scan to remove orphaned meshIDs useful anymore? I'm thinking in context of #282 and #218, that we could hit two birds with one stone by saving the transforms associated with each meshID. However, especially for #218, it would be important to keep them all rather than removing ones that don't contribute triangles.

Reply:
We can remove the scan by using the hash table index as some unique ID for each mesh id, although I am not sure how much of a difference can this make. And this will require a slightly different API. (get raw index or something)
This is probably not related to saving transforms because the hash table is just temporary, we need another thing to store them.

Reply:
Oh right, I forgot we were also using the scan to assign new ids. Thanks!